lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -401,146 +575,231 @@ void AccessConflictAnalysis::identifyBeginAccesses() {
   }
 }
 
-/// When a conflict is detected, flag the access so it can't be folded, and
-/// remove its index from the current access set so we stop checking for
-/// conflicts. Erasing from SparseAccessSet does not invalidate any iterators.
-static void recordConflict(AccessInfo &info, SparseAccessSet &accessSet) {
-  info.setSeenNestedConflict();
-  accessSet.setConflict(info.getAccessIndex());
+void RegionStorageAnalysis::identifyStorageLocations() {
+  for (auto it : result.accessMap) {
+    auto storage = it.second;
+    storageAccesses.insert(storage);
+  }
+}
+
+void RegionStorageAnalysis::calcBottomUpOrder(
+    LoopRegion *region, llvm::SmallVectorImpl<unsigned> &worklist) {
+  worklist.push_back(region->getID());
+  for (auto regionIndex : region->getReverseSubregions()) {
+    auto *region = LRFI->getRegion(regionIndex);
+    if (region->isBlock())
+      continue;
+    calcBottomUpOrder(region, worklist);
+  }
+}
+
+RegionStorageAnalysis::~RegionStorageAnalysis() {
+  // clear the RegionToRegionStorageMap
+  while (!regionToStorageMap.empty()) {
+    auto curr = regionToStorageMap.begin();
+    auto currData = curr->getSecond();
+    regionToStorageMap.erase(curr);
+    delete currData;
+  }
+}
+
+void RegionStorageAnalysis::mergeBBPredAccesses(unsigned id) {
+  auto currStorageIt = regionToStorageMap.find(id);
+  assert(currStorageIt != regionToStorageMap.end() &&
+         "Expected a region storage");
+  auto *storage = currStorageIt->getSecond();
+  auto *region = LRFI->getRegion(id);
+  assert(region->isBlock() && "Expected a block");
+  for (auto pred : region->getPreds()) {
+    auto *predRegion = LRFI->getRegion(pred);
+    if (predRegion->getParentID() != region->getParentID()) {
+      // predecessor out of current sub-region
+      continue;
+    }
+    if (predRegion->isLoop()) {
+      // predecessor out of current sub-region
+      continue;
+    }
+    auto storageIt = regionToStorageMap.find(pred);
+    if (storageIt == regionToStorageMap.end())
+      continue;
+    assert(storageIt != regionToStorageMap.end() &&
+           "Expected sub-region to have been processed");
+    auto *subStorage = storageIt->getSecond();
+    storage->mergeFrom(*subStorage);
+  }
+}
+
+RegionStorage *
+RegionStorageAnalysis::getRegionStorageFromIns(SILInstruction *instr) {
```

Comment:
I'm very concerned that this helper exists and that it needs to be called just to update data flow. It's definitely not the right way to do data flow.